framework/filters/AccessRule.php

```diff
@@ -156,8 +171,13 @@ protected function matchRole($user)
                 if (!$user->getIsGuest()) {
                     return true;
                 }
-            } elseif ($user->can($role)) {
-                return true;
+            } else {
+                $params = empty($this->ruleParamsCallback) ? [] :
```

Comment:
My feedback here: $user->can($role) part could be separated from the guest and 'all' checks, because adding rule checks even with a custom filter is pretty difficult